### DISCUSSION TITLE
《Go 语言设计与实现》 · /golang

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#556] [Announcements]
https://draveness.me/golang/

**Comment by wujunze** [comment]
nice

**Comment by draveness** [comment]
> @wujunze
> nice

你是怎么发现的..

**Comment by yinpeihao** [comment]
> @draveness
> > @wujunze
> > nice
> 
> 你是怎么发现的..

我也发现了

**Comment by draveness** [comment]
> > @draveness
> > > @wujunze
> > > nice
> > 
> > 
> > 你是怎么发现的..
> 
> 我也发现了

昨天开放了...

**Comment by yinpeihao** [comment]
> @draveness
> > > @draveness
> > > > @wujunze
> > > > nice
> > > 
> > > 
> > > 你是怎么发现的..
> > 
> > 我也发现了
> 
> 昨天开放了...

哈哈哈哈，德莱文流皮

**Comment by wujunze** [comment]
@draveness  还有几章要写的    好像没有完全写完啊   😀

**Comment by draveness** [comment]
还有不少要写的😅

**Comment by wujunze** [comment]
> 还有不少要写的😅
期待你的好文章  👍

**Comment by daisuki-konbini** [comment]
有PDF 版本么

**Comment by tczf1128** [comment]
问一下，图是用什么工具画的

**Comment by llussy** [comment]
> @tczf1128
> 问一下，图是用什么工具画的

看到了，是用的Sketch

**Comment by zyjblockchain** [comment]
打个卡，开始精读，感谢作者的分享

**Comment by lily-lee** [comment]
你好，请问有参考资料/文献和延展阅读之类的吗？

**Comment by draveness** [comment]
> 你好，请问有参考资料/文献和延展阅读之类的吗？

暂时没有

**Comment by wxxsw** [comment]
我从 Rails 转到 iOS，又从 iOS 往 Go 转，和你的方向有一点相似，但你实在太厉害了，各方面都能做的这么好，以后要多多拜读你写的东西，向大神看齐！😊

**Comment by beihai0xff** [comment]
打卡,宝藏博主

**Comment by feilengcui008** [comment]
建议明确一下分析的go版本

**Comment by draveness** [comment]
> @feilengcui008
> 建议明确一下分析的go版本

目前是 1.13，会加一章准备工作，里面会介绍当前的 Go 语言版本

**Comment by xiaoheiAh** [comment]
大佬是真的牛皮,疯狂输出

**Comment by beckbikang** [comment]
nice too

**Comment by beckbikang** [comment]
看了大佬的github的commit，神人一枚啊。我想问一句，大佬单身吗？:)

**Comment by draveness** [comment]
> 看了大佬的github的commit，神人一枚啊。我想问一句，大佬单身吗？:)

😃

**Comment by xing00lian** [comment]
大佬，图是用什么画的

**Comment by Charles-Xiao** [comment]
楼主一个人要把这块写的深入浅出感觉耗时会比较久，话说楼主有打算分工协作一下么？以及后续出版一下？

**Comment by draveness** [comment]
> @xing00lian
> 大佬，图是用什么画的

https://draveness.me/sketch-and-sketch

**Comment by draveness** [comment]
> @Charles-Xiao
> 楼主一个人要把这块写的深入浅出感觉耗时会比较久，话说楼主有打算分工协作一下么？以及后续出版一下？

暂时还没有

**Comment by flyfree6265** [comment]
何时出书啊，等着买呢

**Comment by draveness** [comment]
> 何时出书啊，等着买呢

有可能今年下半年

**Comment by ilvcyy** [comment]
这书真的是详解源码，为什么不出书呢？毕竟go已经是十多年的语言了，再不出书它就老了，剩余40%在哪里可以看到？

**Comment by draveness** [comment]
> 这书真的是详解源码，为什么不出书呢？毕竟go已经是十多年的语言了，再不出书它就老了，

也不怕晚，这不还在发展么

> 剩余40%在哪里可以看到？

还在写呢...

**Comment by CiaoTuning** [comment]
期待后续！何时开通付费通道呀？:D 我扫您还是您扫我（大雾）

**Comment by draveness** [comment]
> 期待后续！何时开通付费通道呀？:D 我扫您还是您扫我（大雾）

仅支持 Bitcoin 匿名支付（大雾）

**Comment by ClayAndMore** [comment]
发现宝藏啊，请教下这个博客是用什么实现的

**Comment by draveness** [comment]
> 发现宝藏啊，请教下这个博客是用什么实现的

Hugo-book

**Comment by llifezou** [comment]
不知道咋发现这些武功秘籍的,现在是不想吃饭不想睡觉就想看文章!如果我在北京遇见作者,一定抱着大腿死皮赖脸拜师...

**Comment by Cherrison** [comment]
第一章在哪里啊?

**Comment by draveness** [comment]
> 第一章在哪里啊?

哈哈哈哈，第一章还没写，主要介绍怎么阅读和编译源代码

**Comment by yangwenmai** [comment]
认同作者的一句话：
>理解 Go 语言的发展史是帮助我们深入理解语言的有效途径，我们不应该忽视这一过程。

**Comment by luojiedev** [comment]
再不好好学习，我就40了

**Comment by gabryzhong** [comment]
👀

**Comment by gabryzhong** [comment]
刚刚开始使用go来提供一些webapi的服务，感觉go在设计业务层时过于面向过程了，使得代码段很难拆分干净

**Comment by draveness** [comment]
> 刚刚开始使用go来提供一些webapi的服务，感觉go在设计业务层时过于面向过程了，使得代码段很难拆分干净

『过于面向过程』是指什么呢，Go 语言是可以使用面向对象的方式编程的

**Comment by liaoxuefeng** [comment]
nice

**Comment by davidxiaobj** [comment]
非常感谢，非常感谢，
我发现这是目前最合适我、我最需要的、有基础、更有深度、系统的golang书

**Comment by lupguo** [comment]
感谢分享，向老师学习！可以让自己把Golang看得更加透彻！

**Comment by chenjinxuan** [comment]
谁都不服，就服你！书写的真的棒，非常非常有深度。

**Comment by cafra** [comment]
> @gabryzhong
> 刚刚开始使用go来提供一些webapi的服务，感觉go在设计业务层时过于面向过程了，使得代码段很难拆分干净

面向interface 编程，会是一个很好的软件组织形式

**Comment by tw7613781** [comment]
大佬，你是我前进的明灯～感谢

**Comment by vitoi** [comment]
优秀

**Comment by ROOKIE20570** [comment]
感谢大佬的书！

**Comment by racketprogram** [comment]
理解常见并+发原语 Mutex

多了一個“+”

**Comment by draveness** [comment]
> 理解常见并+发原语 Mutex
> 
> 多了一個“+”

我怎么没找到这个错误

**Comment by racketprogram** [comment]
> @draveness
> > 理解常见并+发原语 Mutex
> > 
> > 多了一個“+”
> 
> 我怎么没找到这个错误

https://ppt.cc/fZWOtx

---

*2020-05-29 UPDATES: 已修复*

**Comment by lgting** [comment]
开始接触学习go语言了， 走到这边是不是我已近成功了一半了

**Comment by ahwgs** [comment]
学习了

**Comment by zhuangzhewei09** [comment]
可以写一些对常用的官方package源码分析啥的嘛

**Comment by draveness** [comment]
> 可以写一些对常用的官方package源码分析啥的嘛

下面不是有标准库么

**Comment by bspp1314** [comment]
博主有出实体书打算吗？

**Comment by draveness** [comment]
> 博主有出实体书打算吗？

会出的

**Comment by xiaolushare** [comment]
膜拜大神, 希望出实体书

**Comment by zhangxu-ai** [comment]
大佬，这本书打算出版吗？一定买来拜读！

**Comment by brilliantguo** [comment]
膜拜大佬，学习了！

**Comment by a1499418300** [comment]
> @yinpeihao
> > @draveness
> > > > @draveness
> > > > > @wujunze
> > > > > nice
> > > > 
> > > > 
> > > > 你是怎么发现的..
> > > 
> > > 我也发现了
> > 
> > 昨天开放了...
> 
> 哈哈哈哈，德莱文流皮

打卡

**Comment by Logic-Luo** [comment]
nice

**Comment by wangsquirrel** [comment]
赞一个👍

**Comment by raven-chen** [comment]
写的真好  手动点赞！

**Comment by hplf** [comment]
这个书针对怎样的群里，貌似不适合入门学习

**Comment by draveness** [comment]
> 这个书针对怎样的群里，貌似不适合入门学习

如果没有编程基础，入门还是看 Go 语言圣经

**Comment by ESWZY** [comment]
虽然刚刚开始学Go，但是我觉得这本书一定会成为高阶程序员必读书！

**Comment by Eminlin** [comment]
很好的文章

**Comment by jhon-von** [comment]
> 大佬，学这个不知道能干啥，是不是学完就知道了呢？

**Comment by dillanzhou** [comment]
大佬，有个疑问，第一章呢？

**Comment by draveness** [comment]
> 大佬，有个疑问，第一章呢？

还没写呢

**Comment by wcxhello2020** [comment]
学习到无法自拔了

**Comment by alpha-baby** [comment]
牛逼

**Comment by Jacob110** [comment]
牛逼，直接从第二章开始，哈哈哈

**Comment by munding** [comment]
比一些卖课的讲的太好了

**Comment by sigma65535** [comment]
> @a1499418300
> > @yinpeihao
> > > @draveness
> > > > > @draveness
> > > > > > @wujunze
> > > > > > nice
> > > > > 
> > > > > 
> > > > > 你是怎么发现的..
> > > > 
> > > > 我也发现了
> > > 
> > > 昨天开放了...
> > 
> > 哈哈哈哈，德莱文流皮
> 
> 打卡

打卡+1

**Comment by so2bin** [comment]
** 写的很不错!**

**Comment by huangwh2014** [comment]
好书啊，我以为有纸质的，结果 没找一以，大神，出版一下啊，赚个辛苦费也可以啊

**Comment by huangwh2014** [comment]
我是刚go初学者，搜索切片与数组区别时就找到你这了，感觉很有用。

**Comment by xiarong715** [comment]
非常好

**Comment by chenlixin93** [comment]
向大佬学习的第一天

**Comment by lipenglin0-0** [comment]
居然没有纸质的
确实是大佬
从内容到境界

**Comment by draveness** [comment]
> 居然没有纸质的

之后会有的

**Comment by handsomeyzz** [comment]
大佬回复我一下噻。看了您的文章我已经从入门到入土了......歪比歪比

**Comment by draveness** [comment]
> 大佬回复我一下噻。看了您的文章我已经从入门到入土了......歪比歪比

回复你的什么？

**Comment by Lightwing-Ng** [comment]
太强了

**Comment by yuchanns** [comment]
前几天在某个群听说这本书就快出实体了！大概什么时候呢？敲碗等一时间购买

**Comment by love477** [comment]
大佬🧍‍♂️

**Comment by Cyan1614** [comment]
大佬有标准库源码解析的计划么

**Comment by draveness** [comment]
> 大佬有标准库源码解析的计划么

最后一章会介绍部分标准库的源代码

**Comment by jiekun** [comment]
Shuqi大佬今年出书的计划有更新消息吗，1月份就看到人邮相关的预告了，不过2020年马上要过去了，想问问大佬实体书的安排有木有落地预计时间，一定支持几本~

**Comment by sunnystitch** [comment]
膜拜大佬  大佬如果出书可以给个签名版么   标准库源码解析什么的什么时候这块的博客有呀

**Comment by icbd** [comment]
作者的博客篇篇精品, 特别是 Golang 系列! 

我在阅读的过程中想用荧光笔做高亮笔记, 一方面方便回顾, 也方便日后搜索, 于是就试着用 Golang 写了一个笔记工具: 

[https://github.com/icbd/gohighlights](https://github.com/icbd/gohighlights)


## Demo

![demo](https://s3.ax1x.com/2020/12/11/rAmv9A.png)


[已经发布到 Chrome Extension Store, 欢迎下载试用](https://chrome.google.com/webstore/detail/go-highlights/homlcfpinafhealhlmjkmdjdejppmmlk)

**Comment by changyonggang** [comment]
> @icbd
> 作者的博客篇篇精品, 特别是 Golang 系列! 
> 
> 我在阅读的过程中想用荧光笔做高亮笔记, 一方面方便回顾, 也方便日后搜索, 于是就试着用 Golang 写了一个笔记工具: 
> 
> [https://github.com/icbd/gohighlights](https://github.com/icbd/gohighlights)
> 
> 
> ## Demo
> 
> ![demo](https://s3.ax1x.com/2020/12/11/rAmv9A.png)
> 
> 
> [已经发布到 Chrome Extension Store, 欢迎下载试用](https://chrome.google.com/webstore/detail/go-highlights/homlcfpinafhealhlmjkmdjdejppmmlk)

大佬，这个插件的server 部署到了哪里？

**Comment by icbd** [comment]
> @changyonggang
> > @icbd
> > 作者的博客篇篇精品, 特别是 Golang 系列! 
> > 
> > 我在阅读的过程中想用荧光笔做高亮笔记, 一方面方便回顾, 也方便日后搜索, 于是就试着用 Golang 写了一个笔记工具: 
> > 
> > [https://github.com/icbd/gohighlights](https://github.com/icbd/gohighlights)
> > 
> > 
> > ## Demo
> > 
> > ![demo](https://s3.ax1x.com/2020/12/11/rAmv9A.png)
> > 
> > 
> > [已经发布到 Chrome Extension Store, 欢迎下载试用](https://chrome.google.com/webstore/detail/go-highlights/homlcfpinafhealhlmjkmdjdejppmmlk)
> 
> 大佬，这个插件的server 部署到了哪里？

Chrome 商店的版本对应的 Server 在腾讯云, 由于域名备案失效了, 现在用的裸 IP😂

推荐使用本地独立部署, 代码开源在这里:

后端:  [https://github.com/icbd/gohighlights](https://github.com/icbd/gohighlights)

前端:  [https://github.com/icbd/gohighlights_ext](https://github.com/icbd/gohighlights_ext)

后端部署依赖 MySQL 和 Elasticsearch, 有问题可以给我发 issue .

**Comment by Dcell** [comment]
看了博主历史文章，这等自我驱动能力让我折服。
我愿称你最强！坐等2020年总结

**Comment by kichijyo** [comment]
您好，我在编译静态库的时候，没有依据包来进行，而是直接针对源文件，进行的创建
就像下面这样:
1.go build -o lib.a src/lib/greet.go
2.go tool compile -I . src/main/main.go
3.go tool link -L .main.o

main…inittask: relocation target lib…inittask not defined

我的源文件：

src/lib/greet.go

package lib
import “fmt”
func Greet(word string){
fmt.Print(word);
}

src/main/main.go

package main
import (
“lib”
)
func main(){
lib.Greet(“Okay”);
}

我想知道，为什么我基于源文件，进行就会报错呢？

**Comment by jwdouble** [comment]
卧槽，大佬超级大佬看了GC和内存管理那几章感觉通透了很多。太强了！

**Comment by Batawar** [comment]
点赞点赞🌚

**Comment by xiaolong302** [comment]
怎么样看全文呢？

**Comment by draveness** [comment]
> 怎么样看全文呢？

左边不是有目录么

**Comment by xiaolong302** [comment]
> @draveness
> > 怎么样看全文呢？
> 
> 左边不是有目录么

不知道 是不是因为COOKIE 的原因， 我并看不到， 不过切换到incognito  模式后看到了。

**Comment by iDvel** [comment]
这个wiki是哪个开源呀，求告知～～

**Comment by draveness** [comment]
> 这个wiki是哪个开源呀，求告知～～

这本书暂时不开源，只有网页版

**Comment by iDvel** [comment]
> @draveness
> > 这个wiki是哪个开源呀，求告知～～
> 
> 这本书暂时不开源，只有网页版

不是这个意思哈，想问下大佬写书的这个网页是用啥生成的呀，好像不是gitbook呀，谢谢～～

**Comment by Aristotle1024** [comment]
建议作者能出视频讲解，更生动形象，这种硬核的讲解看文字的话还是稍显晦涩，不太好理解

**Comment by XiaoGe2030** [comment]
必须打卡 👀

**Comment by LDity** [comment]
大佬什么时候出书啊？想买一本

**Comment by marxlee** [comment]
对此详细的阐述, 我想了解一下, 作者是通过那些资料了解到go的内核信息的呢, 还请作者给些建议做参考

**Comment by draveness** [comment]
> @marxlee
> 对此详细的阐述, 我想了解一下, 作者是通过那些资料了解到go的内核信息的呢, 还请作者给些建议做参考

主要通过 Go 语言的源代码


> @LDity
> 大佬什么时候出书啊？想买一本

应该是下半年


> @EthanYue
> 建议作者能出视频讲解，更生动形象，这种硬核的讲解看文字的话还是稍显晦涩，不太好理解

视频讲解估计是没戏了，太占用时间

**Comment by rsmelody** [comment]
要怎样做，才能像大佬一样，毕业三四年就这么强😅

**Comment by qiushido** [comment]
想问一下作者 纸质书籍出版后，当前Web版还能继续浏览吗？会不会因为出版而关闭当前浏览入口

**Comment by CNife** [comment]
期待大佬的书，到时候一定买一本。

**Comment by haierkeys** [comment]
大佬.. 问一下.. 这个书 有没有离线阅读版本啊.. =.=

**Comment by draveness** [comment]
> @haierspi
> 大佬.. 问一下.. 这个书 有没有离线阅读版本啊.. =.=

短期内应该也不会搞 PDF 版

**Comment by wendao7** [comment]
写作理念小节："大量的篇幅都都是复制的源代码"多了一个"都"字;
主要内容小节："编译原理、运行时、基础知识和进阶知识"建议按照文章整体顺序"编译原理、基础知识、运行时、和进阶知识"

**Comment by auula** [comment]
感谢大佬的书，非常好！

**Comment by syatao** [comment]
👀

**Comment by Changjinxing** [comment]
收藏

**Comment by edte** [comment]
建议还是弄个 pdf，比只是在线要好传播许多

**Comment by jiekun** [comment]
> 建议还是弄个 pdf，比只是在线要好传播许多

马上要出实体书啦，到时候想线下读的可以购买呀；出电子版盗版泛滥的问题就会严重很多，虽然作者本身就开源在个人网站上，但是个人还是挺反对做一套PDF出来的 :(

**Comment by fengbianyun** [comment]
> > 建议还是弄个 pdf，比只是在线要好传播许多
> 
> 马上要出实体书啦，到时候想线下读的可以购买呀；出电子版盗版泛滥的问题就会严重很多，虽然作者本身就开源在个人网站上，但是个人还是挺反对做一套PDF出来的 :(

建议用google搜一波，pdf往上还是挺多的。一直在期待作者的实体书。前段时间终于把这本书追着源码通读了一遍，发现要理解的东西太多了。读起来有点儿吃力，准备再查查资料做笔记，开始深入的干第二遍了。（每篇里面reference建议读一读，收货也不小）

**Comment by cuppar** [comment]
写作理念

分析语言设计与实现的书籍不是特别的好写，很多文章都会陷入对不重要的细节，大量的篇幅**_都都_**是复制的源代码


---

*2021-06-09 UPDATES：已修复*

**Comment by cuppar** [comment]
主要内容

本书的主要内容可以分成四个主要部分，分别是编译原理、运行时、基础知识和进阶知识，这四部分几乎可以覆盖 Go 语言从编译到运行的方方面面，让我们对 Go 语言有更加整体和深刻的认识。下图的**__思维导读__**中展示了本书的主要内容：

---

*2021-06-09 UPDATES：已修复*

**Comment by damonYuan** [comment]
大神， 这个电子书的模板是开源的吗，能不能提供一个链接。

**Comment by Neverlandoo** [comment]
祺总，缺女朋友吗，想和你谈恋爱

**Comment by yizhanggou** [comment]
大佬什么时候出书呀

**Comment by lelelemonade** [comment]
我对这本书的评价是质量及其高，非常有意义

**Comment by ouweibin** [comment]
打卡学习

**Comment by taoyuanyuan** [comment]
大佬牛逼，什么时候出实体书？ 期待

**Comment by WhiteN1ght** [comment]
好好学习，早日滚蛋

**Comment by csDeng** [comment]
遇到大哥的文章，三生有幸

**Comment by chenyufeifei** [comment]
老师，你好。我想请问一下纸质书和博客有什么区别吗？

**Comment by draveness** [comment]
> @chenyufeifei  老师，你好。我想请问一下纸质书和博客有什么区别吗？

https://draveness.me/golang-book-release

**Comment by ThreeBearsDan** [comment]
在评论中输入字母`s`的时候会跳到搜索框里

**Comment by guowei-gong** [comment]
你好，d 哥，我按照书中第 9 页下的底部方式，获取源代码，在浏览器输入：draveness.me/golang/tree/cmd/compile/internal/gc.complieFunctions，没有出现预期的代码，请问是我的理解有误吗？书中内容如下图：

![image](https://user-images.githubusercontent.com/62499904/145680730-3c0f95f5-d7d1-444b-890f-31dd41e66759.png)


在书中的第 7 页，`自下而上或自上而下地扫约`，我理解应该是`扫描`，因为我在 google 搜索：编译器 扫约，并没有这种关键字的页面，大部分都是 `扫描`。书中内容如下图：
![image](https://user-images.githubusercontent.com/62499904/145680712-fd6e3fd9-3c3d-4b42-b905-776c4b5b2e59.png)

**Comment by draveness** [comment]
> 你好，d 哥，我按照书中第 9 页下的底部方式，获取源代码，在浏览器输入：draveness.me/golang/tree/cmd/compile/internal/gc.complieFunctions，没有出现预期的代码，请问是我的理解有误吗？书中内容如下图：
> 
> 在书中的第 7 页，`自下而上或自上而下地扫约`，我理解应该是`扫描`，因为我在 google 搜索：编译器 扫约，并没有这种关键字的页面，大部分都是 `扫描`。书中内容如下图： 

相关的信息已经同步给编辑老师

**Comment by dablelv** [comment]
博主，请问该站点是怎么搭建的？

**Reply by lnnt** [reply]
https://github.com/draveness/blog-comments/issues/161#issuecomment-613782282

**Comment by XiaoGe2030** [comment]
邮件已收到！

**Comment by XiaoGe2030** [comment]
邮件已收到！
